File: android/core/src/main/kotlin/com/slipstream/core/discovery/SubnetSweepStrategy.kt
package com.slipstream.core.discovery

import com.slipstream.core.SlipstreamPorts
import com.slipstream.core.net.LocalNetwork
import com.slipstream.core.net.SubnetMath
import java.net.InetSocketAddress
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit

/**
 * S4: parallel sweep of every host in the local /24, probing each on the control port.
 * Spec §5: backstop for access points that drop multicast. Bounded to a /24 by
 * [SubnetMath.enumerateHosts] (empty for anything wider), and to [maxConcurrentProbes]
 * simultaneous probes by a [Semaphore] so a sweep never opens 65k sockets at once.
 */
class SubnetSweepStrategy(
    private val probe: PeerProbe,
    private val maxConcurrentProbes: Int = 254,
) : DiscoveryStrategy {

    override val name: String = "subnet-sweep"

    override suspend fun find(network: LocalNetwork): DiscoveredPeer? = coroutineScope {
        val hosts = SubnetMath.enumerateHosts(network.localAddress, network.prefixLength).toList()
        if (hosts.isEmpty()) return@coroutineScope null

        val semaphore = Semaphore(maxConcurrentProbes)
        val result = CompletableDeferred<DiscoveredPeer?>()

        val jobs = hosts.map { host ->
            launch {
                semaphore.withPermit {
                    if (result.isCompleted) return@withPermit
                    val found = try {
                        probe.probe(InetSocketAddress(host, SlipstreamPorts.CONTROL))
                    } catch (e: Exception) {
                        null
                    }
                    if (found != null) {
                        result.complete(found)
                    }
                }
            }
        }

        launch {
            jobs.joinAll()
            result.complete(null)
        }

        val found = result.await()
        jobs.forEach { it.cancel() }
        found
    }
}

File: windows/src/Slipstream.Core/Discovery/SubnetSweepStrategy.cs
using System.Net;
using Slipstream.Core.Net;

namespace Slipstream.Core.Discovery;

/// <summary>
/// Spec §5 S4. The backstop for access points that silently drop multicast.
/// Bounded to a /24 by SubnetMath — a wider sweep is refused, not attempted.
/// </summary>
public sealed class SubnetSweepStrategy(PeerProbe probe, int maxConcurrency = 254) : IDiscoveryStrategy
{
    public string Name => "subnet-sweep";

    public async Task<DiscoveredPeer?> FindAsync(LocalNetwork network, CancellationToken cancellationToken)
    {
        var hosts = SubnetMath
            .EnumerateHosts(network.LocalAddress, network.PrefixLength)
            .ToList();

        if (hosts.Count == 0) return null;

        using var linked = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        using var slots = new SemaphoreSlim(maxConcurrency);

        DiscoveredPeer? winner = null;

        var probes = hosts.Select(async host =>
        {
            await slots.WaitAsync(linked.Token);
            try
            {
                var found = await probe(new IPEndPoint(host, SlipstreamPorts.Control), linked.Token);
                if (found is not null)
                {
                    Interlocked.CompareExchange(ref winner, found, null);
                    await linked.CancelAsync();
                }
            }
            catch (OperationCanceledException)
            {
                // Another probe won, or the caller cancelled.
            }
            finally
            {
                slots.Release();
            }
        });

        try
        {
            await Task.WhenAll(probes);
        }
        catch (OperationCanceledException)
        {
            // Expected once a winner cancels the rest.
        }

        return winner;
    }
}
